graylog2-server/src/main/java/org/graylog2/plugin/Message.java

```diff
@@ -318,16 +361,65 @@ private void addField(final String key, final Object value, final boolean isRequ
         }
 
         if (FIELD_TIMESTAMP.equals(trimmedKey) && value != null && value instanceof Date) {
-            fields.put(FIELD_TIMESTAMP, new DateTime(value));
+            final DateTime timestamp = new DateTime(value);
+            final Object previousValue = fields.put(FIELD_TIMESTAMP, timestamp);
+            updateSize(trimmedKey, timestamp, previousValue);
         } else if (value instanceof String) {
             final String str = ((String) value).trim();
 
             if (isRequiredField || !str.isEmpty()) {
-                fields.put(trimmedKey, str);
+                final Object previousValue = fields.put(trimmedKey, str);
+                updateSize(trimmedKey, str, previousValue);
             }
         } else if (value != null) {
-            fields.put(trimmedKey, value);
+            final Object previousValue = fields.put(trimmedKey, value);
+            updateSize(trimmedKey, value, previousValue);
+        }
+    }
+
+    private void updateSize(String fieldName, Object newValue, Object previousValue) {
```

Comment:
The `setSource()` method sets the new source value directly in the `fields` map and thus doesn't do the size counting via `updateSize()`.